saga/saga.go

```diff
@@ -1,146 +1,197 @@
-	return Saga{
-		log: log,
-	}
+// Concurrent Object Representing a Saga
+// Methods update the state of the saga or
+// Provide access to the Current State
+type Saga struct {
+	id       string
+	log      SagaLog
+	state    *SagaState
+	updateCh chan sagaUpdate
+	mutex    *sync.RWMutex
 }
 
-/*
- * Start Saga. Logs Message message to the log.
- * Returns the resulting SagaState or an error if it fails.
- */
-func (s Saga) StartSaga(sagaId string, job []byte) (*SagaState, error) {
+// Start a New Saga.  Logs a Start Saga Message to the SagaLog
+// returns a Saga, or an error if one occurs
+func newSaga(sagaId string, job []byte, log SagaLog) (*Saga, error) {
 
-	//Create new SagaState
 	state, err := makeSagaState(sagaId, job)
 	if err != nil {
 		return nil, err
 	}
 
-	//Durably Store that we Created a new Saga
-	err = s.log.StartSaga(sagaId, job)
+	err = log.StartSaga(sagaId, job)
 	if err != nil {
 		return nil, err
 	}
 
-	return state, nil
+	updateCh := make(chan sagaUpdate, 0)
+
+	s := &Saga{
+		id:       sagaId,
+		log:      log,
+		state:    state,
+		updateCh: updateCh,
+		mutex:    &sync.RWMutex{},
+	}
+
+	go s.updateSagaStateLoop()
+
+	return s, nil
+}
+
+// Rehydrate a saga from a specified SagaState, does not write
+// to SagaLog assumes that this is a recovered saga.
+func rehydrateSaga(sagaId string, state *SagaState, log SagaLog) *Saga {
+	updateCh := make(chan sagaUpdate, 0)
+	s := &Saga{
+		id:       sagaId,
+		log:      log,
+		state:    state,
+		updateCh: updateCh,
+		mutex:    &sync.RWMutex{},
+	}
+
+	if !state.IsSagaCompleted() {
+		go s.updateSagaStateLoop()
+	}
+
+	return s
 }
 
-/*
- * Log an End Saga Message to the log, returns updated SagaState
- * Returns the resulting SagaState or an error if it fails
- */
-func (s Saga) EndSaga(state *SagaState) (*SagaState, error) {
-	return s.logMessage(state, MakeEndSagaMessage(state.sagaId))
+// Returns the Current Saga State
+func (s *Saga) GetState() *SagaState {
+	s.mutex.RLock()
+	defer s.mutex.RUnlock()
+	return s.state
 }
 
-/*
- * Log an AbortSaga message.  This indicates that the
- * Saga has failed and all execution should be stopped
- * and compensating transactions should be applied.
- *
- * Returns the resulting SagaState or an error if it fails
- */
-func (s Saga) AbortSaga(state *SagaState) (*SagaState, error) {
+//
+// Log an End Saga Message to the log, returns updated SagaState
+// Returns the resulting SagaState or an error if it fails
+//
+// Once EndSaga is succesfully called, trying to log additional
+// messages will result in a panic.
+func (s *Saga) EndSaga() error {
+	return s.updateSagaState(MakeEndSagaMessage(s.id))
+}
 
-	return s.logMessage(state, MakeAbortSagaMessage(state.sagaId))
+//
+// Log an AbortSaga message.  This indicates that the
+// Saga has failed and all execution should be stopped
+// and compensating transactions should be applied.
+//
+// Returns an error if it fails
+//
+func (s *Saga) AbortSaga() error {
```

Comment:
(I was reading this code and answered my own question: this should not close. After you write an abort message, you have to start writing begin comp task messages, and then end comp task messages, and only after all that the close. So yes, you have to log more messages after abort'ing.)